pennylane/optimize/rotosolve.py

```diff
@@ -14,78 +14,122 @@
-        array[int]: List of flattened inputs, with each input being unwrapped if it has length 1
-    """
-    if num_freqs is None:
-        num_freqs = [1] * sum(requires_grad)
-        num_freqs_flat = num_freqs
-    elif np.isscalar(num_freqs):
-        _assert_integer(num_freqs)
-        num_freqs = [num_freqs] * sum(requires_grad)
-        num_freqs_flat = num_freqs
-    else:
-        num_freqs_flat = []
-        for num_frequency in num_freqs:
-            if np.isscalar(num_frequency):
-                _assert_integer(num_frequency)
-                num_frequency_flat = num_frequency
-            else:
-                num_frequency_flat = list(_flatten(num_frequency))
-                for _num_freq in num_frequency_flat:
-                    _assert_integer(_num_freq)
-                num_frequency_flat = np.array(num_frequency_flat, dtype=int)
-            num_freqs_flat.append(num_frequency_flat)
-        if len(num_freqs) != sum(requires_grad):
-            raise ValueError(
-                "The length of the provided numbers of frequencies "
-                f"({len(num_freqs)}) does not match the number of function arguments "
-                f"({sum(requires_grad)})."
-            )
-    return num_freqs, num_freqs_flat
+import pennylane as qml
 
 
 def _brute_optimizer(fun, num_steps, **kwargs):
     r"""Brute force optimizer, wrapper of scipy.optimizer.brute that repeats it
-    ``num_steps`` times. Signature is as expected by ``RotosolveOptimizer._rotosolve``
-    below, providing a scalar minimal position and the function value at that position."""
+    ``num_steps`` times. Signature is as expected by ``RotosolveOptimizer._min_numeric``
+    below, returning a scalar minimal position and the function value at that position."""
     width = 2 * np.pi
     x_min = 0.0
     Ns = kwargs.pop("Ns")
     for _ in range(num_steps):
         _range = (x_min - width / 2, x_min + width / 2)
         x_min, y_min, *_ = brute(fun, ranges=(_range,), full_output=True, Ns=Ns, **kwargs)
+        # We only ever use this function for 1D optimization
+        x_min = x_min[0]
         width /= Ns
 
     return x_min, y_min
 
 
 def _shgo_optimizer(fun, **kwargs):
     r"""Wrapper for ``scipy.optimize.shgo`` (Simplicial Homology global optimizer).
-    Signature is as expected by ``RotosolveOptimizer._rotosolve`` below, providing
+    Signature is as expected by ``RotosolveOptimizer._min_numeric`` below, providing
     a scalar minimal position and the function value at that position."""
     opt_res = shgo(fun, **kwargs)
-    return opt_res.x, opt_res.fun
+    return opt_res.x[0], opt_res.fun
+
+
+def _prepare_optimizer(optimizer, optimizer_kwargs):
+    r"""Set default optimizer and optimizer keyword arguments
+    for the one-dimensional optimization in each substep of Rotosolve."""
+    optimizer_kwargs = optimizer_kwargs or {}
+    optimizer = optimizer or "brute"
```

Comment:
I wonder if this is simplified by using default argument values? e.g.,
```suggestion
def _prepare_optimizer(optimizer="brute", optimizer_kwargs=None):
    r"""Set default optimizer and optimizer keyword arguments
    for the one-dimensional optimization in each substep of Rotosolve."""
    optimizer_kwargs = optimizer_kwargs or {}
```

this is very minor though, doesn't need to be accepted if it adds complication further down.

Reply:
I think the current implementation is a bit clearer on what's happening: No defaults are set anyway up in the pipeline, but only at the `_prepare_optimizer` level are they set. This makes sense because the default `optimizer_kwargs` depend on the value of `optimizer`, I think. One could indeed implement this differently but it seems a bit clearer as is.
What do you think?

Reply:
Happy to leave it as is!